testing-guidelines/ci.md

```diff
@@ -0,0 +1,118 @@
+
+- __python checks__ - whenever python source, dependencies or config change.
+- __javascript checks__ - whenever javascript source, dependencies or config change.
+- __functional and visual checks__ - whenever any sopurce of config changes (most of the time).
+- __repo hygiene checks__ - always.
+
+Checks almost always run when the CI config has changed.
+
+If a check can be skipped, the status is set to `success` (green tick) to satisfy the GitHub required checks, but the message will have a text of `Skipped`.
+
+
+### Quality
+
+We run a series of quality checks on the repo. These range from static checks like linting to unit tests all the way through to fully end-to-end functional tests. 
+
+All tests have a name of something like `test-<type>-<os>-<stability-level>`. `os` and `stability-level` are optional.
+
+This is a simple breakdown of our current quality checks:
+
+| Language   | Check           | operating system | Workflow file            | Notes |
+| ---------- | --------------- | ---------------- | ------------------------ | ----- |
+| Python     | Linting         | linux            | `test-python.yml`        |       |
+| Python     | Formatting      | linux            | `test-python.yml`        |       |
+| Python     | Type-checking   | linux            | `test-python.yml`        |       |
+| Python     | Unit tests      | linux            | `test-python.yml`        |       |
+| Python     | Unit tests      | windows          | `test-python.yml`        |       |
+| JavaScript | Linting         | linux            | `test-js.yml`            |       |
+| JavaScript | Formatting      | linux            | `test-js.yml`            |       |
+| JavaScript | Type-checking   | linux            | `test-js.yml`            |       |
+| JavaScript | Unit tests      | linux            | `test-js.yml`            |       |
+| n/a        | Functional      | linux            | `test-functional/yml`    |       |
+| n/a        | Visual          | linux            | `deploy+test-visual/yml` |       |
+| n/a        | Large files     | linux            | `test-hygiene.yml`       | Checks that all files are below 5 MB |
+| n/a        | Notebooks match | linux            | `test-hygiene.yml`       | Ensures that notebooks and demos are in sync |
+
+
+One important thing to note is that we split 'flaky' and 'non-flaky' python unit/ integration test out. These test are flaky because of network requests that they make, they are typically fine but anything that can cause a red check in PRs makes us less trustworthy of our Ci and confidence is the goal! The windows test are also very slow and only test a few edge cases. The flaky and windows tests are not run in every PR but are always run against the release PR to ensure everything is working as expected prior to a release. All other checks are run for every pull request, ensuring everything will work when we merge into `main`.
```

Comment:
```suggestion
One important thing to note is that we split 'flaky' and 'non-flaky' Python unit/integration tests out. These tests are flaky because of network requests that they make. They are typically fine, but anything that can cause a red check in PRs makes us less trustworthy of our CI and confidence is the goal! The Windows tests are also very slow and only test a few edge cases. The flaky and windows tests are not run in every PR, but are always run against the release PR to ensure everything is working as expected prior to a release. All other checks are run for every pull request, ensuring everything will work when we merge into `main`.
```